src/sql/workbench/parts/query/browser/queryActions.ts

```diff
@@ -415,6 +416,65 @@ export class ListDatabasesAction extends QueryTaskbarAction {
 	}
 }
 
+/**
+ * Action class that toggles SQLCMD mode for the editor
+ */
+export class ToggleSqlCmdModeAction extends QueryTaskbarAction {
+
+	public static EnableSqlcmdClass = 'enablesqlcmd';
+	public static DisableSqlcmdClass = 'disablesqlcmd';
+	public static ID = 'ToggleSqlCmdModeAction';
+
+	private _enablesqlcmdLabel = nls.localize('enablesqlcmdLabel', "Enable SQLCMD Mode");
+	private _disablesqlcmdLabel = nls.localize('disablesqlcmdLabel', "Disable SQLCMD Mode");
+	constructor(
+		editor: QueryEditor,
+		private _isSqlCmdMode: boolean,
+		@IQueryManagementService protected readonly queryManagementService: IQueryManagementService,
+		@IConfigurationService protected readonly configurationService: IConfigurationService,
+		@IConnectionManagementService connectionManagementService: IConnectionManagementService
+	) {
+		super(connectionManagementService, editor, ToggleSqlCmdModeAction.ID, undefined);
+	}
+
+	public get isSqlCmdMode(): boolean {
+		return this._isSqlCmdMode;
+	}
+
+	public set isSqlCmdMode(value: boolean) {
+		this._isSqlCmdMode = value;
+		this.updateLabelAndIcon();
+	}
```

Comment:
It seems weird to have this state stored in the action. That seems like it should be tied to the editor directly instead and this just handles telling the various parts that care about it what the new value is

Reply:
Hey, this is following the same pattern as other buttons... see **ToggleConnectDatabaseAction** 

(the reasoning is that the editor switch calls constructor with state value every time)

Reply:
Hrmmm...I still don't think it's appropriate (and would argue the same thing for ToggleConnectDatabaseAction). You're already storing the state in the editor state anyways so it just seems like duplication of effort here and could potentially cause syncing issues if later on we wanted to add this action somewhere else. 

@anthonydresser Do you have any thoughts on this since you've been doing a lot of work with actions/query editor state?